src/core/desktopentrymonitor.cpp

```diff
@@ -0,0 +1,188 @@
+#include "desktopentrymonitor.hpp"
+
+#include <qdir.h>
+#include <qdiriterator.h>
+#include <qfileinfo.h>
+#include <qlogging.h>
+#include <qloggingcategory.h>
+#include <qtenvironmentvariables.h>
+
+#include "desktoputils.hpp"
+#include "logcat.hpp"
+
+namespace {
+QS_LOGGING_CATEGORY(logDesktopMonitor, "quickshell.desktopentrymonitor", QtWarningMsg);
+}
+
+DesktopEntryMonitor::DesktopEntryMonitor(QObject* parent): QObject(parent) {
+	this->watcher = new QFileSystemWatcher(this);
+	this->debounceTimer = new QTimer(this);
+	this->debounceTimer->setSingleShot(true);
+	this->debounceTimer->setInterval(500); // 500ms debounce
+
+	// Initialize XDG desktop paths
+	this->initializeDesktopPaths();
+
+	// Setup connections
+	connect(
+	    this->watcher,
+	    &QFileSystemWatcher::directoryChanged,
+	    this,
+	    &DesktopEntryMonitor::onDirectoryChanged
+	);
+	connect(
+	    this->watcher,
+	    &QFileSystemWatcher::fileChanged,
+	    this,
+	    &DesktopEntryMonitor::onFileChanged
+	);
+	connect(this->debounceTimer, &QTimer::timeout, this, &DesktopEntryMonitor::processChanges);
+
+	// Start monitoring
+	this->startMonitoring();
+}
+
+void DesktopEntryMonitor::initializeDesktopPaths() {
+	this->desktopPaths = DesktopUtils::getDesktopDirectories();
+}
+
+void DesktopEntryMonitor::startMonitoring() {
+	for (const QString& path: this->desktopPaths) {
+		if (QDir(path).exists()) {
+			qCDebug(logDesktopMonitor) << "Monitoring desktop entry path:" << path;
+			this->addDirectoryHierarchy(path);
+		}
+	}
+}
+
+void DesktopEntryMonitor::addDirectoryHierarchy(const QString& dirPath) {
+	this->scanAndWatch(dirPath);
+}
+
+void DesktopEntryMonitor::scanAndWatch(const QString& dirPath) {
```

Comment:
scanAndWatch + all the various change handlers share a lot of the code. consider initialization to be a diff from an empty set of entries and share the code.